docs/WritingTests.asciidoc

```diff
@@ -22,52 +22,207 @@ https://github.com/os-autoinst/openQA[official repository].
+    # the bootloader screen will timeout itself
+    assert_screen "bootloader", 15;
+
+    # press enter to boot right away
+    send_key "ret";
+
+    # wait for the desktop to appear
+    assert_screen "desktop", 300;
+}
+-------------------------------------------------------------------
+
+`assert_screen` & `send_key` are provided by {testapi}.
+
+==== `test_flags`
+
+Specifies what should happen when test execution of the current test module is
+finished depending on the result.
+
+Each flag is defined with a hash key, the possible hash keys are:
+
+* *fatal*: When set to `1` the whole test suite is aborted if the test module
+   fails. The overall state is set to `failed`.
+* *ignore_failure*: When set to `1` and the test module fails, it will not
+   affect the overall result at all.
+* *milestone*: After this test succeeds, update the 'lastgood' snapshot of the
+   SUT.
+* *no_rollback*: Don't roll back to the 'lastgood' snapshot of the SUT if the
+   test module fails.
+* *always_rollback*: Roll back to the 'lastgood' snapshot of the SUT even if
+   test was successful.
+
+See the example below for how to enable a test flag. Note that snapshots are
+only supported by the QEMU backend. When using other backends `fatal` is
+therefore enabled by default. One can explicitly set it to `0` to disable the
+behavior for all backends even though it is not possible to roll back.
+
+An example usage:
+
+[source,perl]
+-------------------------------------------------------------------
+sub test_flags {
+    return {fatal => 1};
+}
+-------------------------------------------------------------------
+
+==== `pre_run_hook`
+Is called before the run function - mainly useful for a whole group of tests.
+It's useful to setup the start point of the test.
+
+An example usage:
+
+[source,perl]
+-------------------------------------------------------------------
+sub pre_run_hook {
+    # Make sure to begin the test in the root console.
+    select_console 'root-console';
+}
+-------------------------------------------------------------------
+
+==== `post_fail_hook`
+
+Is called only when the test execution failed, it's useful to upload log files
+or determine the state of the machine.
+
+An example usage:
+
+[source,perl]
+-------------------------------------------------------------------
+sub post_fail_hook {
+    # Take an screenshot when the test failed
+    save_screenshot;
+}
+-------------------------------------------------------------------
+
+
+==== `post_run_hook`
+
+Is run after successful run function, it will be executed regardless of the
+result of the test run.
+
+An example usage:
+
+[source,perl]
+-------------------------------------------------------------------
+sub post_fail_hook {
+    send_key 'ctrl-alt-f3';
+
+    assert_script_run 'openqa-cli api experimental/search q=shutdown.pm' ;    
+}
+-------------------------------------------------------------------
+
+=== Example Perl Test Modules
+[id="testmodule_perl_examples"]
```

Comment:
Hm, I switched the lines for the `[id=` and the `===` and didn't affect at all the id generation. it marks the h3 with that ID regardless of the position.

Also the headers above do it like that too.

What do you mean by "following a reference"?

Reply:
> Also the headers above do it like that too.

We're trying to improve the documentation for years but it isn't perfect yet.

> What do you mean by "following a reference"?

Clicking on an internal cross reference within the document or following a URL with hash like https://open.qa/docs/#snapshots. Just have a look at this URL. When loading this page you'll end up below the actual heading. It would be much nicer if the heading was still visible which is achieved by moving the ID before the heading. Of course one could also just use the auto-generated ID (and only that is the ID that marks the h3 element, see e.g. https://open.qa/docs/#_using_snapshots_to_speed_up_development_of_tests).

Reply:
Huh, I see now.

It looks like a deeper problem with the CSS Layout, normally the referred id should be the first item in the viewport. I don't quite know rn how to improve it without going way off-track with CSS changes or adding workarounds in the asciidocs.

Reply:
> normally the referred id should be the first item in the viewport.

It should be but that's also not the problem here. If you use the `[id="…"]` you'll get the ID on the next paragraph. So the paragraph ends up as the first item in the viewport and not the heading. Maybe the HTML generator behaves not like that in all cases (e.g. when the element after the heading is no paragraph but something else) but it makes still generally sense to put the ID before the heading.

Reply:
I guess if you wanted to change the ID of the heading element itself then you would write it like this:

```
[[usersguide]]
= openQA users guide
```

Reply:
> It should be but that's also not the problem here. If you use the [id="…"] you'll get the ID on the next paragraph.

Just checked this, and it doesn't render like that. IIUC this:

```asciidoc
=== Example Perl Test Modules
[id="testmodule_perl_examples"]
```

is expected to render:

```html
<!-- [...] -->
<h3>Example Perl test modules</h3>
<!-- the next element in line has the ID -->
<div id="_example_perl_test_modules"> class="paragraph">
<!-- [...] -->
```

But it actually renders:

```html
<!-- [...] -->
<h3 id="_example_perl_test_modules">Example Perl test modules</h3>
<div id="testmodule_perl_examples" class="paragraph">
<!-- [...] -->
```

HTML seems to be rendered correctly with `[id=...]` below the heading.